### PR TITLE
Integrate Best Practices for a More Robust and Maintainable Code Base

### PR DESCRIPTION
**Pull request by ashwinprasadme**
Similar to the Scalpel pull request, this PR aims to improve the overall quality and maintainability of the project by refactoring the code base and implementing common best practices. The following changes have been made:

1. **Linting:** Added `ruff`, `isort`, and `black` as linting tools to ensure that the code adheres to PEP8 standards. Configured the tools to run automatically via pre-commit hooks. Furthermore, all files inside the `pycg` directory are now linted.

2. **Pre-commit:** Installed pre-commit to manage pre-commit hooks. Configured hooks to run linting tasks automatically before committing changes.

3. **Setup Script:** Created `setup.sh` and `requirements.txt` to manage the development environment including dependencies, install pre-commit hooks, and streamline the installation process of the development environment for contributors. This will help ensure that all developers work with the same package versions and minimize issues related to package version discrepancies.

I have run both the micro-benchmark and unit tests and the results are not altered after refactoring.

---

Furthermore, the following bugs are fixed in the ground truth of the micro-benchmark:

+ For test `dicts_test.DictsTest.test_assign` the ground truth is wrong. func1 is never called.
+ For test `dicts_test.DictsTest.test_nested` the ground truth is wrong. func1 is never called.
+ For test `dicts_test.DictsTest.test_update` the ground truth is wrong. func1 is never called.

---

Also, one open question, is there a Python version that we should target going forward? I would suggest we officially test and support Python 3.10

Please review the changes and let me know if you have any suggestions or concerns.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        name: isort (pyi)
+        types: [pyi]
+
+-   repo: https://github.com/psf/black
+    rev: 23.1.0
+    hooks:
+    -   id: black
+ci:
+    autofix_commit_msg: 🎨 [pre-commit.ci] Auto format from pre-commit.com hooks
+    autoupdate_commit_msg: ⬆ [pre-commit.ci] pre-commit autoupdate
```
Comment (vitsalis) [inline .pre-commit-config.yaml:43]: Not very experienced about pre-commit hooks, but from my understanding, those hooks run on my local machine when I perform certain actions. I do not feel particularly safe with running scripts retrieved from remote entities (`ruff-pre-commit` has less than 200 stars) on my local machine each time.

File: setup_devenv.sh [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    ${PYTHON} -m venv .env
+    if [ $? -ne 0 ]; then
+        echo "Could not create virtual environment. Leaving now"
+        exit 1
+    fi
+    updateenv
+}
+
+function install() {
+
```
Comment (vitsalis) [inline setup_devenv.sh:116]: whiteline

File: setup_devenv.sh [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    echo_block "Installing mandatory dependencies"
+
+    if [ -x "$(command -v apt-get)" ]; then
+        echo "Debian/Ubuntu detected. Setup for this system in-progress"
+        install_debian
+    else
+        echo "This script does not support your OS."
+        echo "If you have Python version 3.8 - 3.10, pip, virtualenv you can continue."
```
Comment (vitsalis) [inline setup_devenv.sh:124]: Above we say that we support Python 3.10, but here we say `3.8 - 3.10`

**Review by vitsalis — COMMENTED** [review]
Thanks for the great work! However, I feel slightly uneasy about the installation and usage of external scripts on my local machine whenever I create a commit. Typically, in the repositories I work with I use a linter as a GitHub Action/CircleCI to identify potential linting issues with the code I create with a pull request.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (ashwinprasadme) [reply .pre-commit-config.yaml:43]: Indeed, your concern is valid. However, this is maintained by the official team and part of their documentation. Check here: https://beta.ruff.rs/docs/usage/#pre-commit

Comment (vitsalis) [reply .pre-commit-config.yaml:43]: Still, a workspace can contain sensitive data and this is a very valid attack vector. We can merge the changes to the files, but for me pre-commit hooks based on external dependencies are a no-go. I would be much more comfortable with a Github actions linter.

**Comment by gdrosos** [comment]
Thanks, @ashwinprasadme  for the awesome work you've put into this! 🎉👏 
We really appreciate your efforts and the time you've invested.

We've made a slight change in our approach and decided to utilize a GitHub Action with established linters instead. I've submitted a new pull request (#57) to implement this.

Your changes to the files are still valuable, and we'd like to merge them into the codebase. Let's make sure we incorporate your work and keep the project moving forward.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (vitsalis) [reply .pre-commit-config.yaml:43]: After an offline discussion with @gdrosos , we decided to introduce GitHub Actions to the repository (see #57 ). So, moving forward with this PR, I would suggest we remove the pre-commit stuff and merge the linting of the files that this PR produces. Thanks a bunch for your hard work on this! More specifically:
- remove the pre-commit config
- Remove pre-commit from `requirements.txt`
- Remove pre-commit stuff from the install script

Comment (ashwinprasadme) [reply .pre-commit-config.yaml]: Sounds great! Let's do that.

I have now removed `pre-commit` and renamed `requirements.txt` to `requirements-dev.txt` with flake8, black, and isort.

**Review by vitsalis — APPROVED** [review]
Thanks!

**Pull request merged**